iquip/apps/dataviewer.py

```diff
@@ -936,6 +1007,29 @@ def setDataset(
         self.policy = SimpleScanDataPolicy(dataset, parameters, units)
         self.frame.sourceWidget.setParameters(parameters, units)
 
+    @pyqtSlot(list)
+    def modifyDataset(self, modifications: List[Dict[str, Any]]):
+        """Modifies the dataset and updates the plot.
+
+        Args:
+            See _DatasetFetcherThread.modified signal.
+        """
+        # TODO(kangz12345@snu.ac.kr): Implement modifications other than "append".
+        if self.policy is None:
+            logger.error("Tried to modify data when data policy is None.")
+            return
+        appended = np.vstack(tuple(m["x"] for m in modifications if m["action"] == "append"))
+        if self.policy.dataset.size == 0:
+            self.policy.dataset = appended
+        else:
+            self.policy.dataset = np.concatenate((self.policy.dataset, appended))
+        if not self.axis:
+            return
+        self.updateMainPlot(self.axis, self.frame.dataPointWidget.dataType())
+        self.thread.mutex.lock()
```

Comment:
Why is it necessary? Is it already locked?

Reply:
This will prevent the wait condition from notifying before the fetcher thread start waiting.

Reply:
Oh, I see! Your implementation is correct.